### ISSUE TITLE
improved HyperLogLog estimator

### ISSUE THREAD

**Issue by StefanKarpinski** [joshday/OnlineStats.jl#177]
See https://arxiv.org/pdf/1706.07290.pdf. The advantage is that it is unbiased across the entire range from small to large cardinality, so it avoids the need to switch to the "linear estimator" for small cardinalities (there is a large spike in variance of the classic HyperLogLog estimator around this cross-over point). Reference implementation (by me, should be checked but seems to work):
```jl
using Random

const m = 2^16
const q = 64

"Generate a sample of unique values"
sample() = rand(1:m), 1 + min(q, leading_zeros(rand(UInt64)))
sample(n::Int) = [sample() for _=1:n]

"Produce a mergeable sketch of a sample"
function sketch(sample::Vector{<:Tuple{Integer,Integer}})
    sk = zeros(Int, m)
    for (register, leading_digit) in sample
        sk[register] = max(sk[register], leading_digit)
    end
    return sk
end

"Compute the sufficient statistics of a sketch"
function stats(sketch::Vector{<:Integer})
    st = Dict{Int,Int}()
    for leading_digit in sketch
        st[leading_digit] = get(st, leading_digit, 0) + 1
    end
    return sort!(collect(st))
end

"Compute the sufficient statistics of a sample"
stats(sample::Vector{<:Tuple{Integer,Integer}}) = stats(sketch(sample))

# functions required to compute the improved estimator

const ∞ = 10^6 # max iterations

function σ(x::Float64)
    y = x
    for k = 1:∞
        y′ = y + x^2^k*exp2(k-1)
        y′ == y && return y
        y = y′
    end
end

function τ(x::Float64)
    y = 1 - x
    x2k = 1/x
    for k = 1:∞
        x2k ^= 2
        y′ = y - (1 - x^0.5^k)^2*exp2(-k)
        y′ == y && return y/3
        y = y′
    end
end

# limit of the bias correction factor α_m
const α_∞ = 1/(2log(2))

# precomputed first and third terms of the denominator:
const T₁ = [m*σ(c/m) for c = 1:m]
const T₃ = [m*τ(1-c/m)*exp2(-q) for c = 1:m]

"The improved HyperLogLog estimator"
function hll(stats::Vector{Pair{T,S}}) where {T<:Integer, S<:Integer}
    d = 0.0
    # uses k, c to match 
    # k: leading digit
    # C_k: register count
    for (k, C_k) in stats
        d += k == 0   ? T₁[C_k] :
             k == q+1 ? T₃[C_k] : C_k*exp2(-k)
    end
    return α_∞*m^2 / d
end

hll(sketch::Vector{<:Integer}) = hll(stats(sketch))
hll(sample::Vector{<:Tuple{Integer,Integer}}) = hll(sketch(sample))
```
The main `hll` estimator method takes in the "sufficient statistics" of the HyperLogLog sketch, which is the count of "registers" which have a given largest first digit index value. This is not mergeable since it discards which buckets have which largest first digit index (it's even less data than a sketch is), but it's all you need to compute the HyperLogLog estimate and it's how the paper expresses the estimator. Example usage of this code:
```jl
julia> n = 10^6

julia> S = sample(n)
1000000-element Array{Tuple{Int64,Int64},1}:
 (53494, 1)
 (11521, 1)
 (48920, 1)
 ⋮
 (24535, 1)
 (63780, 1)
 (33809, 2)

julia> sketch(S)
65536-element Array{Int64,1}:
  4
  5
  6
  3
  4
  ⋮
  4
  3
  9
  2
  5

julia> stats(S)
21-element Array{Pair{Int64,Int64},1}:
  1 => 27
  2 => 1399
  3 => 8295
  4 => 15668
  5 => 15323
  6 => 10985
  7 => 6468
  8 => 3633
  9 => 1819
 10 => 929
 11 => 475
 12 => 237
 13 => 147
 14 => 65
 15 => 33
 16 => 19
 17 => 7
 18 => 4
 19 => 1
 20 => 1
 21 => 1

julia> hll(S)
999615.7076733122

julia> abs(hll(S) - n)/n
0.00038429232668783515
```

**Comment by joshday** [comment]
Cool!  I'll check it out.

**Comment by StefanKarpinski** [comment]
It's a really well written paper and I tried to stick to the paper's notation as much as I could. Hopefully it should be pretty clear.

**Comment by brucala** [comment]
I guess this can be closed after #213
